src/librustc_lint/unused.rs

```diff
@@ -464,3 +464,61 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnusedAllocation {
         }
     }
 }
+
+declare_lint! {
+    pub(super) UNUSED_LABEL,
+    Warn,
+    "warns on unused labels"
+}
+
+#[derive(Clone)]
+pub struct UnusedLabel(pub Vec<ast::Label>);
+
+impl UnusedLabel {
+    pub fn new() -> Self {
+        UnusedLabel(vec![])
+    }
+}
+
+impl LintPass for UnusedLabel {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(UNUSED_LABEL)
+    }
+}
+
+impl EarlyLintPass for UnusedLabel {
+    fn check_expr(&mut self, _: &EarlyContext, expr: &ast::Expr) {
+        match expr.node {
+            ast::ExprKind::While(_, _, Some(label))
+            | ast::ExprKind::WhileLet(_, _, _, Some(label))
+            | ast::ExprKind::ForLoop(_, _, _, Some(label))
+            | ast::ExprKind::Loop(_, Some(label)) => {
+                self.0.push(label);
```

Comment:
Can you add a comment along the lines of "adding all labels as we encounter their definitions"